nixos/modules/services/development/mlflow/default.nix

```diff
@@ -0,0 +1,132 @@
+    else if v == true then
+      keyToFlag k
+    else if lib.isList v then
+      lib.concatMapStringsSep " " (x: toFlags k x) v
+    else
+      "${keyToFlag k} ${toString v}";
+
+  flags = lib.concatStringsSep " " (lib.mapAttrsToList toFlags cfg.settings);
+
+  valueType = with lib.types;
+    nullOr (oneOf [ bool float int package path port str ]);
+
+  settingsType = with lib.types;
+    submodule {
+      freeformType = attrsOf (oneOf [ valueType (listOf valueType) ]);
+      options = {
+        host = lib.mkOption {
+          type = str;
+          default = "127.0.0.1";
+          description = ''
+            The IP address the mlflow server should listen on.
+          '';
+        };
+
+        port = lib.mkOption {
+          type = port;
+          default = 5000;
+          description = ''
+            The TCP port the mlflow server should listen on.
+          '';
+        };
+
+        default-artifact-root = lib.mkOption {
+          type = str;
+          default = if cfg.settings.serve-artifacts or false then
+            "mlflow-artifacts:/"
+          else
+            "/var/lib/mlflow/mlruns";
+
+          defaultText = lib.literalExpression ''
+            if cfg.settings.serve-artifacts or false then
+              "mlflow-artifacts:/"
+            else
+              "/var/lib/mlflow/mlruns"
+          '';
+
+          description = ''
+            Directory in which to store artifacts for any new experiments
+            created. For tracking server backends that rely on SQL, this option
+            is required in order to store artifacts.
+          '';
+        };
+
+        backend-store-uri = lib.mkOption {
+          type = str;
+          default = "sqlite:////var/lib/mlflow/mlruns.db";
+          description = ''
+            URI to which to persist experiment and run data.
+
+            Acceptable URIs are SQLAlchemy-compatible database connection
+            strings (e.g. 'sqlite:///path/to/file.db') or local filesystem URIs
+            (e.g. 'file:///absolute/path/to/directory'). Using a SQL-backed
+            store is required for a number of features.
+          '';
+        };
+      };
+    };
+
+in {
+  options.services.mlflow = with lib; {
+    enable = mkEnableOption "the MLflow tracking server";
+
+    settings = mkOption {
+      type = settingsType;
+      default = { };
+      description = ''
+        Settings for mlflow server, passed to it as flags.
+      '';
+
+      example = literalExpression ''
+        {
+          serve-artifacts = true;
+        }
+      '';
+    };
+
+    package = mkOption {
+      type = types.package;
+      default = pkgs.python3.pkgs.mlflow;
+      defaultText = "pkgs.python3.pkgs.mlflow";
+      description = ''
+        The mlflow derivation to use.
+      '';
+    };
+
+    autoUpgrade = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Automatically upgrade the mlflow backend store to new schema versions.
```

Comment:
I think the service simply fails to start.

Reply:
Plus, in this context, people using PostgreSQL or something server-based will have to do `systemd.services.mlflow.after = [ "postgresql.service" ]` to avoid failures at runtime.

Reply:
Wouldn't it be better to follow the `stateVersion` mechanism here? I do not know how much `mlflow` will have many major versions, but following NextCloud / PGSQL ways, we could pin the version and let the user do an explicit unpinning.
Otherwise, what would happen if we run mlflow v(N + 1) on a database having data for version N without upgrading, does the service fails to start?

Reply:
As far as I remember the service will fail to start and a sysadmin can manually run the migration or downgrade mlflow or whatever. If we coupled the schema version to the `stateVersion`, we'd need to support many versions of mlflow. For postgres, that's easy because it's popular but mlflow isn't that popular so I think it's better to keep it simple?